lib/wibox/layout/grid.lua

```diff
@@ -0,0 +1,484 @@
+    end
+
+    return false
+end
+
+function grid:set(index, widget2)
+    if (not widget2) or (not self._private.rows[index]) then return false end
+
+    base.check_widget(widget2)
+
+    self._private.rows[index] = widget2
+
+    self:emit_signal("widget::layout_changed")
+
+    return true
+end
+
+--- Insert a new widget in the layout at position `index`
+-- @tparam number index The position
+-- @param widget The widget
+-- @treturn boolean If the operation is successful
+function grid:insert(index, widget)
+    local TODO_TOTAL = math.huge --FIXME
+    if not index or index < 1 or index > TODO_TOTAL then return false end
+
+    base.check_widget(widget)
+    table.insert(self._private.rows, index, widget) --FIXME
+    self:emit_signal("widget::layout_changed")
+
+    return true
+end
+
+-- Fit the grid layout into the given space
+-- @param context The context in which we are fit.
+-- @param orig_width The available width.
+-- @param orig_height The available height.
+function grid:fit(context, orig_width, orig_height)
+    local max_col, max_row = get_cell_sizes(self, context, orig_width, orig_height)
+
+    -- Now that all fit is done, get the maximum
+    local used_max_h, used_max_w = 0, 0
+
+    for row_idx = 1, #max_row do
+        -- The other widgets will be discarded
+        if used_max_h + max_row[row_idx] > orig_height then
+            break
+        end
+
+        used_max_h = used_max_h + max_row[row_idx]
+    end
+
+    for col_idx = 1, #max_col do
+        -- The other widgets will be discarded
+        if used_max_w + max_col[col_idx] > orig_width then
+            break
+        end
+
+        used_max_w = used_max_w + max_col[col_idx]
+    end
+
+    return used_max_w, used_max_h
+end
+
+function grid:reset()
+    self._private.rows = {}
+    self._private.children = {}
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Set the layout's fill_space property. If true, the remaining space will
+-- be distributed equally across all widgets in the grid.
+-- @DOC_wibox_layout_grid_fill_space_EXAMPLE@
+-- @property fill_space
+-- @param boolean
+
+function grid:set_fill_space(val)
+    if self._private.fill_space ~= val then
+        self._private.fill_space = not not val
+        self:emit_signal("widget::layout_changed")
+    end
+end
+
+--- The the border (grid) width.
+-- @DOC_wibox_layout_grid_border_EXAMPLE@
+-- @property border_width
+-- @param number
+
+function grid:set_border_width(val)
+    if self._private.border_width ~= val then
+        self._private.border_width = math.max(0, val)
+
+        -- This count as spacing, so the layout really change
+        self:emit_signal("widget::layout_changed")
+    end
+end
+
+--- The the border (grid) color.
+-- @property border_color
+-- @param color
+-- @see gears.color
+
+function grid:set_border_color(val)
+    self._private.border_color = color(val)
+
+    self:emit_signal("widget::redraw_needed")
+end
+
+local function get_layout(dir, widget1, ...)
```

Comment:
I guess you saw what I used for base at this point. I should cleanup my template.